File: web/src/geo/geo.ts
/**
 * Geodesy helpers. Distances in meters, angles in degrees unless noted.
 * Headings are compass bearings: 0 = north, clockwise positive.
 */

export type LonLat = [number, number]

export interface Pose {
  lon: number
  lat: number
  ele: number
  /** compass heading in degrees, 0 = north, clockwise */
  heading: number
  /** camera pitch in degrees, 0 = level, positive looking up */
  pitch: number
  /** horizontal accuracy in meters */
  accuracy: number
  speedMps: number
  ts: number
}

export const EARTH_R = 6371000
const D2R = Math.PI / 180
const R2D = 180 / Math.PI

export const clamp = (v: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, v))
export const lerp = (a: number, b: number, t: number) => a + (b - a) * t
export const wrap360 = (deg: number) => ((deg % 360) + 360) % 360

/** Signed smallest rotation from `from` to `to`, in [-180, 180). */
export const angDiffDeg = (from: number, to: number) => ((to - from + 540) % 360) - 180

/** Interpolate compass angles along the shortest arc. */
export const lerpAngleDeg = (a: number, b: number, t: number) => wrap360(a + angDiffDeg(a, b) * t)

export function haversineM(a: LonLat, b: LonLat): number {
  const p1 = a[1] * D2R
  const p2 = b[1] * D2R
  const dp = p2 - p1
  const dl = (b[0] - a[0]) * D2R
  const s = Math.sin(dp / 2) ** 2 + Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2
  return 2 * EARTH_R * Math.asin(Math.sqrt(s))
}

export function bearingDeg(a: LonLat, b: LonLat): number {
  const p1 = a[1] * D2R
  const p2 = b[1] * D2R
  const dl = (b[0] - a[0]) * D2R
  const y = Math.sin(dl) * Math.cos(p2)
  const x = Math.cos(p1) * Math.sin(p2) - Math.sin(p1) * Math.cos(p2) * Math.cos(dl)
  return wrap360(Math.atan2(y, x) * R2D)
}

export function destPoint(a: LonLat, brgDeg: number, distM: number): LonLat {
  const br = brgDeg * D2R
  const d = distM / EARTH_R
  const p1 = a[1] * D2R
  const l1 = a[0] * D2R
  const p2 = Math.asin(Math.sin(p1) * Math.cos(d) + Math.cos(p1) * Math.sin(d) * Math.cos(br))
  const l2 =
    l1 + Math.atan2(Math.sin(br) * Math.sin(d) * Math.cos(p1), Math.cos(d) - Math.sin(p1) * Math.sin(p2))
  return [wrap360(l2 * R2D + 180) - 180, p2 * R2D]
}

/**
 * Flat local frame around a reference point. Good to well under 0.1% error
 * at region scale (tens of km). x = meters east, y = meters north.
 */
export class LocalFrame {
  readonly kx: number
  readonly ky: number
  constructor(
    readonly lon0: number,
    readonly lat0: number,
  ) {
    this.kx = 111320 * Math.cos(lat0 * D2R)
    this.ky = 110574
  }
  toXY(lon: number, lat: number): [number, number] {
    return [(lon - this.lon0) * this.kx, (lat - this.lat0) * this.ky]
  }
  toLonLat(x: number, y: number): LonLat {
    return [this.lon0 + x / this.kx, this.lat0 + y / this.ky]
  }
}

export interface XY {
  x: number
  y: number
}

/** Project p onto segment ab. Returns clamped t and squared distance. */
export function projectOnSegment(
  px: number,
  py: number,
  ax: number,
  ay: number,
  bx: number,
  by: number,
): { t: number; x: number; y: number; d2: number } {
  const dx = bx - ax
  const dy = by - ay
  const seg2 = dx * dx + dy * dy
  let t = 0
  if (seg2 > 0) t = clamp(((px - ax) * dx + (py - ay) * dy) / seg2, 0, 1)
  const x = ax + t * dx
  const y = ay + t * dy
  const ex = px - x
  const ey = py - y
  return { t, x, y, d2: ex * ex + ey * ey }
}

/** Douglas-Peucker simplification on planar points (units = tol units). */
export function simplifyDP<T extends XY>(pts: T[], tol: number): T[] {
  if (pts.length < 3) return pts.slice()
  const keep = new Uint8Array(pts.length)
  keep[0] = keep[pts.length - 1] = 1
  const stack: Array<[number, number]> = [[0, pts.length - 1]]
  const tol2 = tol * tol
  while (stack.length) {
    const [i0, i1] = stack.pop()!
    let dmax = -1
    let imax = -1
    for (let i = i0 + 1; i < i1; i++) {
      const pr = projectOnSegment(pts[i].x, pts[i].y, pts[i0].x, pts[i0].y, pts[i1].x, pts[i1].y)
      if (pr.d2 > dmax) {
        dmax = pr.d2
        imax = i
      }
    }
    if (dmax > tol2 && imax > 0) {
      keep[imax] = 1
      stack.push([i0, imax], [imax, i1])
    }
  }
  const out: T[] = []
  for (let i = 0; i < pts.length; i++) if (keep[i]) out.push(pts[i])
  return out
}

/** 16-wind compass name for a bearing. */
export function compass16(deg: number): string {
  const names = [
    'N', 'NNE', 'NE', 'ENE', 'E', 'ESE', 'SE', 'SSE',
    'S', 'SSW', 'SW', 'WSW', 'W', 'WNW', 'NW', 'NNW',
  ]
  return names[Math.round(wrap360(deg) / 22.5) % 16]
}

File: ios/TrailSight/Core/Geo.swift
//
//  Geo.swift
//  TrailSight
//
//  Geodesy helpers, a direct port of the web app's geo.ts so both clients
//  agree on every number. Distances are meters, angles are degrees unless
//  noted. Headings are compass bearings: 0 = north, clockwise positive.
//

import Foundation

/// Mean Earth radius in meters, matching the web engine.
public let earthRadiusM: Double = 6_371_000

private let d2r = Double.pi / 180
private let r2d = 180 / Double.pi

/// A geographic coordinate. `lon` is degrees east, `lat` degrees north.
public struct LonLat: Equatable, Sendable {
    public var lon: Double
    public var lat: Double

    public init(lon: Double, lat: Double) {
        self.lon = lon
        self.lat = lat
    }

    /// Builds from a `[lon, lat, ...]` JSON array slice, the format used by
    /// region packs. Returns nil if fewer than two elements are present.
    public init?(array: [Double]) {
        guard array.count >= 2 else { return nil }
        self.init(lon: array[0], lat: array[1])
    }
}

/// A fused position and orientation sample used by the nav engine and HUD.
public struct Pose: Sendable {
    public var lon: Double
    public var lat: Double
    /// Elevation in meters above sea level.
    public var ele: Double
    /// Compass heading in degrees, 0 = north, clockwise.
    public var heading: Double
    /// Camera pitch in degrees, 0 = level, positive looking up.
    public var pitch: Double
    /// Horizontal accuracy in meters.
    public var accuracy: Double
    public var speedMps: Double
    /// Timestamp in seconds since reference date.
    public var ts: Double

    public init(
        lon: Double, lat: Double, ele: Double, heading: Double,
        pitch: Double = 0, accuracy: Double = 5, speedMps: Double = 0, ts: Double = 0
    ) {
        self.lon = lon
        self.lat = lat
        self.ele = ele
        self.heading = heading
        self.pitch = pitch
        self.accuracy = accuracy
        self.speedMps = speedMps
        self.ts = ts
    }

    public var lonLat: LonLat { LonLat(lon: lon, lat: lat) }
}

/// Clamps `v` into `[lo, hi]`.
public func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double {
    min(hi, max(lo, v))
}

/// Linear interpolation between `a` and `b`.
public func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
    a + (b - a) * t
}

/// Wraps any angle into `[0, 360)`.
public func wrap360(_ deg: Double) -> Double {
    let m = deg.truncatingRemainder(dividingBy: 360)
    return (m + 360).truncatingRemainder(dividingBy: 360)
}

/// Signed smallest rotation from `from` to `to`, in `[-180, 180)`.
public func angDiffDeg(from: Double, to: Double) -> Double {
    let raw = (to - from + 540).truncatingRemainder(dividingBy: 360)
    return (raw + 360).truncatingRemainder(dividingBy: 360) - 180
}

/// Interpolates compass angles along the shortest arc.
public func lerpAngleDeg(_ a: Double, _ b: Double, _ t: Double) -> Double {
    wrap360(a + angDiffDeg(from: a, to: b) * t)
}

/// Great-circle distance between two coordinates in meters (haversine).
public func haversineM(_ a: LonLat, _ b: LonLat) -> Double {
    let p1 = a.lat * d2r
    let p2 = b.lat * d2r
    let dp = p2 - p1
    let dl = (b.lon - a.lon) * d2r
    let sinDp = sin(dp / 2)
    let sinDl = sin(dl / 2)
    let s = sinDp * sinDp + cos(p1) * cos(p2) * sinDl * sinDl
    return 2 * earthRadiusM * asin(sqrt(s))
}

/// Initial compass bearing from `a` to `b`, degrees in `[0, 360)`.
public func bearingDeg(from a: LonLat, to b: LonLat) -> Double {
    let p1 = a.lat * d2r
    let p2 = b.lat * d2r
    let dl = (b.lon - a.lon) * d2r
    let y = sin(dl) * cos(p2)
    let x = cos(p1) * sin(p2) - sin(p1) * cos(p2) * cos(dl)
    return wrap360(atan2(y, x) * r2d)
}

/// Destination point given a start, a bearing, and a distance in meters.
public func destPoint(from a: LonLat, bearingDeg brg: Double, distanceM: Double) -> LonLat {
    let br = brg * d2r
    let d = distanceM / earthRadiusM
    let p1 = a.lat * d2r
    let l1 = a.lon * d2r
    let p2 = asin(sin(p1) * cos(d) + cos(p1) * sin(d) * cos(br))
    let l2 = l1 + atan2(sin(br) * sin(d) * cos(p1), cos(d) - sin(p1) * sin(p2))
    return LonLat(lon: wrap360(l2 * r2d + 180) - 180, lat: p2 * r2d)
}

/// Flat local East-North frame around a reference point. Good to well under
/// 0.1 percent error at region scale (tens of km). x = meters east of the
/// origin, y = meters north of the origin. This is the same equirectangular
/// approximation the web engine uses, and it is also the horizontal part of
/// the ENU frame the AR layer places content in.
public struct LocalFrame: Sendable {
    public let lon0: Double
    public let lat0: Double
    /// Meters per degree of longitude at the origin latitude.
    public let kx: Double
    /// Meters per degree of latitude.
    public let ky: Double

    public init(lon0: Double, lat0: Double) {
        self.lon0 = lon0
        self.lat0 = lat0
        self.kx = 111_320 * cos(lat0 * d2r)
        self.ky = 110_574
    }

    /// Projects a coordinate to local meters (x east, y north).
    public func toXY(lon: Double, lat: Double) -> (x: Double, y: Double) {
        ((lon - lon0) * kx, (lat - lat0) * ky)
    }

    public func toXY(_ p: LonLat) -> (x: Double, y: Double) {
        toXY(lon: p.lon, lat: p.lat)
    }

    /// Inverse of `toXY`.
    public func toLonLat(x: Double, y: Double) -> LonLat {
        LonLat(lon: lon0 + x / kx, lat: lat0 + y / ky)
    }
}

/// Result of projecting a point onto a segment: clamped parameter `t`,
/// the closest point, and the squared distance to it.
public struct SegmentProjection: Sendable {
    public let t: Double
    public let x: Double
    public let y: Double
    public let d2: Double
}

/// Projects point `(px, py)` onto segment `(ax, ay)-(bx, by)` in the plane.
public func projectOnSegment(
    px: Double, py: Double,
    ax: Double, ay: Double,
    bx: Double, by: Double
) -> SegmentProjection {
    let dx = bx - ax
    let dy = by - ay
    let seg2 = dx * dx + dy * dy
    var t = 0.0
    if seg2 > 0 {
        t = clamp(((px - ax) * dx + (py - ay) * dy) / seg2, 0, 1)
    }
    let x = ax + t * dx
    let y = ay + t * dy
    let ex = px - x
    let ey = py - y
    return SegmentProjection(t: t, x: x, y: y, d2: ex * ex + ey * ey)
}

/// 16-wind compass name for a bearing, e.g. 22 degrees -> "NNE".
public func compass16(_ deg: Double) -> String {
    let names = [
        "N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE",
        "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW",
    ]
    let idx = Int((wrap360(deg) / 22.5).rounded()) % 16
    return names[idx]
}
